### PR TITLE
Support y_pred as pandas DataFrame for surrogate model predictions

### PR DESCRIPTION
**Pull request by gaugup**
Seems like if y_pred is pandas dataframe, then slicing the teacher model predictions has issues in MimicExplainer. Handling this scenario by using iloc for slicing.

Signed-off-by: Gaurav Gupta <gaugup@microsoft.com>

### CODE REVIEW

File: python/interpret_community/mimic/model_distill.py [hunk 1]
```diff
@@ -76,7 +77,10 @@ def _model_distill(teacher_model_predict_fn, uninitialized_surrogate_model, data
         surrogate_model = uninitialized_surrogate_model(**explainable_model_args)
     if is_classifier and teacher_y.shape[1] == 2:
         # Make sure output has only 1 dimension
-        teacher_y = teacher_y[:, 1]
+        if isinstance(teacher_y, pd.DataFrame):
```
Comment (imatiach-msft) [inline python/interpret_community/mimic/model_distill.py:80]: the teacher_y can be a DataFrame?  In that case, I would recommend to just change this code:
```
teacher_y = teacher_model_predict_fn(original_training_data)
```
to convert teacher_y to numpy right after that line.
Otherwise, the multiclass surrogate code will probably break as well.

File: test/models.py [hunk 2]
```diff
@@ -67,3 +67,28 @@ def predict_proba(self, X_pred):
...
+        pass
+
+    def fit(self, X, y):
+        pass
+
+    def predict(self, X_pred):
+        result = np.repeat(0, X_pred.shape[0])
+        result[0] = 1
+        if self.return_predictions_as_dataframe:
+            return pd.DataFrame(result)
```
Comment (imatiach-msft) [inline test/models.py:84]: crazy to see this output a dataframe as the prediction

File: test/test_mimic_explainer.py [hunk 3]
```diff
@@ -295,6 +295,15 @@ def test_dataframe_model(self, mimic_explainer, sample_cnt_per_grain, grains_dic
             kwargs['categorical_features'] = ['fruit']
         mimic_explainer(model, X, LGBMExplainableModel, features=features, model_task=model_task, **kwargs)
 
+    @pytest.mark.parametrize('if_predictions_as_dataframe', [True, False])
+    def test_explain_model_binary_classification_with_different_format_predictions(
```
Comment (imatiach-msft) [inline test/test_mimic_explainer.py:299]: please add test for multiclass surrogate model as well, as the code there may break on this output type as well.
Probably a good idea to add a test for regression models too, although it seems like that might be safer - not sure if all surrogate model types support dataframe as labels

File: test/test_mimic_explainer.py [hunk 4]
```diff
@@ -295,6 +295,15 @@ def test_dataframe_model(self, mimic_explainer, sample_cnt_per_grain, grains_dic
...
+            self, mimic_explainer, if_predictions_as_dataframe):
+        x_train, y_train, X_test, y_test, classes = create_binary_classification_dataset()
+        model = PredictAsDataFrameTestModel(return_predictions_as_dataframe=if_predictions_as_dataframe)
+        model.fit(x_train, y_train)
+        kwargs = {'explainable_model_args': {'n_jobs': 1}, 'augment_data': False, 'reset_index': True}
+        mimic_explainer(model, x_train, LGBMExplainableModel, **kwargs)
```
Comment (imatiach-msft) [inline test/test_mimic_explainer.py:305]: LightGBM handles a lot more types than scikit-learn usually - I would just loop over all surrogate model types, since I'm not sure if the scikit-learn ones can actually handle the labels column as a dataframe -- unless we decide to convert it to numpy which I think might be more manageable

**Review by imatiach-msft — APPROVED** [review]

File: python/interpret_community/mimic/model_distill.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 """Utilities to train a surrogate model from teacher."""
```
Comment (imatiach-msft) [inline python/interpret_community/mimic/model_distill.py:5]: thinking about this more, we shouldn't handle this in model_distill function but in the model wrapper.  We can have a wrapper for models that output pandas dataframes for predict/predict_proba.

Comment (gaugup) [reply python/interpret_community/mimic/model_distill.py:5]: Thanks for the suggestion. I am wrapping the models giving dataframe as outputs for predict/predict_proba and converting the outputs into numpy arrays

File: test/test_mimic_explainer.py (hunk 4 above)
Comment (gaugup) [reply test/test_mimic_explainer.py:305]: Done.

File: test/models.py (hunk 2 above)
Comment (gaugup) [reply test/models.py:84]: I know. But it is good to cover this case.

File: python/interpret_community/common/model_wrapper.py [hunk 6]
```diff
@@ -344,6 +370,14 @@ def _classifier_without_proba(model):
     return isinstance(model, SGDClassifier) and not hasattr(model, SKLearn.PREDICT_PROBA)
 
 
+def _does_predict_gives_dataframe(model, examples):
+    import pandas as pd
```
Comment (imatiach-msft) [inline python/interpret_community/common/model_wrapper.py:374]: nit: move imports to top of file

File: python/interpret_community/common/model_wrapper.py [hunk 7]
```diff
@@ -242,6 +242,30 @@ def predict(self, dataset):
         return self._eval_function(dataset)
 
 
+class WrappedDataFramePRedictModel(object):
```
Comment (imatiach-msft) [inline python/interpret_community/common/model_wrapper.py:245]: nit: PRedictModel -> PredictModel
any reason R capital?

File: python/interpret_community/common/model_wrapper.py [hunk 8]
```diff
@@ -242,6 +242,30 @@ def predict(self, dataset):
...
+    """A class for wrapping a model which give pandas DataFrame as outputs of predictions."""
+
+    def __init__(self, model):
+        """Initialize the WrappedDataFramePRedictModel with the model function."""
```
Comment (imatiach-msft) [inline python/interpret_community/common/model_wrapper.py:249]: nit: same here (WrappedDataFramePRedictModel)

File: python/interpret_community/common/model_wrapper.py [hunk 9]
```diff
@@ -344,6 +370,14 @@ def _classifier_without_proba(model):
     return isinstance(model, SGDClassifier) and not hasattr(model, SKLearn.PREDICT_PROBA)
 
 
+def _does_predict_gives_dataframe(model, examples):
+    import pandas as pd
+    if isinstance(model.predict(examples.dataset), pd.DataFrame) or \
```
Comment (imatiach-msft) [inline python/interpret_community/common/model_wrapper.py:375]: calling predict on full dataset is a bit wasteful, I think this can either be added in _eval_function method, after:

result = function(examples.typed_wrapper_func(examples_dataset[0]))

or we can keep this and just use something similar to _eval_model and only select first instance:

result = model.predict_proba(examples.typed_wrapper_func(examples.dataset[0:1]))

I think either approach works fine, maybe second one is easier and also better to make

File: python/interpret_community/mimic/model_distill.py [hunk 10]
```diff
@@ -5,6 +5,7 @@
...
 
 import numpy as np
+import pandas as pd
```
Comment (imatiach-msft) [inline python/interpret_community/mimic/model_distill.py:8]: nit: is this import needed still?

**Review by imatiach-msft — APPROVED** [review]

File: python/interpret_community/mimic/model_distill.py (hunk 10 above)
Comment (gaugup) [reply python/interpret_community/mimic/model_distill.py:8]: Done.

File: python/interpret_community/mimic/model_distill.py (hunk 5 above)
Comment (gaugup) [reply python/interpret_community/mimic/model_distill.py]: Changed to handle the inside the wrapped models for classification and regression.

File: python/interpret_community/common/model_wrapper.py (hunk 9 above)
Comment (gaugup) [reply python/interpret_community/common/model_wrapper.py:359]: Moved the logic inside the classification and regression wrapped classes. Please have a look.

File: python/interpret_community/common/model_wrapper.py (hunk 8 above)
Comment (gaugup) [reply python/interpret_community/common/model_wrapper.py]: Removed.

File: python/interpret_community/common/model_wrapper.py (hunk 7 above)
Comment (gaugup) [reply python/interpret_community/common/model_wrapper.py]: Removed.

File: test/test_mimic_explainer.py (hunk 3 above)
Comment (gaugup) [reply test/test_mimic_explainer.py]: Done.

File: python/interpret_community/mimic/model_distill.py (hunk 1 above)
Comment (gaugup) [reply python/interpret_community/mimic/model_distill.py]: fixed

File: python/interpret_community/common/model_wrapper.py (hunk 6 above)
Comment (gaugup) [reply python/interpret_community/common/model_wrapper.py]: done

File: python/interpret_community/common/model_wrapper.py [hunk 11]
```diff
@@ -205,6 +205,8 @@ def predict(self, dataset):
         if is_sequential or isinstance(self._model, WrappedPytorchModel):
             return self._model.predict_classes(dataset).flatten()
         preds = self._model.predict(dataset)
+        if isinstance(preds, pd.DataFrame):
+            preds = preds.values.ravel()
```
[the remark this answers is not in the document]
Comment (gaugup) [reply python/interpret_community/common/model_wrapper.py:209]: Yes ravel was needed in case of regression problems. the .values on pandas dataframe returns a 2-d numpy array with shape like (len, ). So ravel is to make that a linear array.

File: python/interpret_community/common/model_wrapper.py [hunk 12]
```diff
@@ -205,6 +205,8 @@ def predict(self, dataset):
         if is_sequential or isinstance(self._model, WrappedPytorchModel):
```
[the remark this answers is not in the document]
Comment (gaugup) [reply python/interpret_community/common/model_wrapper.py:205]: I think we very less testing around prediction function in interpret-community. Created issue https://github.com/interpretml/interpret-community/issues/380.
